### ISSUE TITLE
[Feature Request] Implement GRP-Obliteration

### ISSUE THREAD

**Issue by Heliumrich** [p-e-w/heretic#184]
GRP-Obliteration is a new unalignment method that uses Group Relative Policy Optimization (a reinforcement-learning-style approach) to remove safety constraints from aligned LLMs with just one unlabeled prompt, reliably making the model comply with harmful requests while keeping its general capabilities mostly intact.

Compared to earlier abliteration techniques (which surgically remove a single "refusal direction" in activations and often need harmful pair data), GRP-Oblit is superior because it achieves stronger and more consistent unalignment (higher average attack success rates, e.g., 81% effectiveness vs. 69% for abliteration), preserves utility better with less degradation on benchmarks like MMLU, requires far less data (one prompt vs. curated datasets), and shows lower variance across many models and architectures.

Paper : https://arxiv.org/pdf/2602.06258
Paper source : https://arxiv.org/abs/2602.06258

**Comment by kabachuha** [comment]
@Heliumrich GRPO requires model training, meaning spawning the gradients and the optimizer, as well as doing multiple rollouts, increasing VRAM consumption in 2x+ times, making it unsuitable for consumer hardware and even for enterprise servers, if we think about abliterating large models such as GLM or Kimi, whereas heretic / ablitration has **no** such overhead.

**Comment by kabachuha** [comment]
Ah yes, and a reference model for KL regularization, so +1 x model VRAM

TL;DR: useless method.

**Comment by kabachuha** [comment]
@Heliumrich If you'd like to increase model's quality and attack success rate, I'd recommend looking at https://github.com/pralab/som-refusal-directions.

They train a very small auxiliary neural network on pre-collected activations, identifying refusal  manifold from which they then select multiple core clusters and use them as abliteration directions, resulting in more precise model modification and counteracting the more advanced recent models where the refusal circuits are more fine grained. Unlike your method, this can be easily integrated into heretic, because the directions acquisition is an in-between pre-processing steps after the hidden states have been acquired and the optimization search is launched.

I opened an issue here #140.

**Comment by Heliumrich** [comment]
Thank you, I don't know much about this, glad you are giving me info instead of saying I am stupid, refreshing for github lol
<3

I don't know if it's truly that much better (less quality decrease) than your technique, but yeah the VRAM usage and finetuning required is a big price to pay

**Comment by p-e-w** [comment]
Yeah, training-based methods are out of scope for Heretic.
